dlt/resource_dlt_pipeline.go

```diff
@@ -0,0 +1,357 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+// DefaultTimeout is the default amount of time that Terraform will wait when creating, updating and deleting pipelines.
+const DefaultTimeout = 20 * time.Minute
+
+type createPipelineRequestStruct struct {
+	pipelines.CreatePipeline
+}
+
+type updatePipelineRequestStruct struct {
+	pipelines.EditPipeline
+}
+
+// Name not keeping twice
+type Pipeline struct {
+	pipelines.PipelineSpec
+	PipelineId          string                              `json:"pipeline_id,omitempty"`
+	AllowDuplicateNames bool                                `json:"allow_duplicate_names,omitempty"`
+	DryRun              bool                                `json:"dry_run,omitempty"`
+	Cause               string                              `json:"cause,omitempty"`
+	ClusterId           string                              `json:"cluster_id,omitempty"`
+	CreatorUserName     string                              `json:"creator_user_name,omitempty"`
+	Health              pipelines.GetPipelineResponseHealth `json:"health,omitempty"`
+	LastModified        int64                               `json:"last_modified,omitempty"`
+	LatestUpdates       []pipelines.UpdateStateInfo         `json:"latest_updates,omitempty"`
+	RunAsUserName       string                              `json:"run_as_user_name,omitempty"`
+	State               pipelines.PipelineState             `json:"state,omitempty"`
+	URL                 string                              `json:"url,omitempty"`
+}
+
+// Constants for PipelineStates
+const (
+	StateDeploying  pipelines.PipelineState = "DEPLOYING"
+	StateStarting   pipelines.PipelineState = "STARTING"
+	StateRunning    pipelines.PipelineState = "RUNNING"
+	StateStopping   pipelines.PipelineState = "STOPPPING"
+	StateDeleted    pipelines.PipelineState = "DELETED"
+	StateRecovering pipelines.PipelineState = "RECOVERING"
+	StateFailed     pipelines.PipelineState = "FAILED"
+	StateResetting  pipelines.PipelineState = "RESETTING"
+	StateIdle       pipelines.PipelineState = "IDLE"
+)
+
+func adjustForceSendFields(clusterList *[]pipelines.PipelineCluster) {
+	for i := range *clusterList {
+		cluster := &((*clusterList)[i])
+		// TF Go SDK doesn't differentiate between the default and not set values.
+		// If nothing is specified, DLT creates a cluster with enhanced autoscaling
+		// from 1 to 5 nodes, which is different than sending a request for zero workers.
+		// The solution here is to look for the Spark configuration to determine
+		// if the user only wants a single node cluster (only master, no workers).
+		if cluster.SparkConf["spark.databricks.cluster.profile"] == "singleNode" {
+			cluster.ForceSendFields = append(cluster.ForceSendFields, "NumWorkers")
+		}
+	}
+}
+
+func (w workspaceClient) Create(ctx context.Context, d *schema.ResourceData, timeout time.Duration) error {
+	var createPipelineRequest createPipelineRequestStruct
```

Comment:
Don't create the createPipelineRequestStruct. Use the Go `CreatePipeline` directly. Add any necessary field and/or alias to the `Pipeline` you define above.

Reply:
I think when we do DataToStructPointer we also need Aliases set for the struct that we want to fill. That's why I created a separate struct for create response and created aliases for it

Reply:
You can do it wihout creating a struct:
https://github.com/databricks/terraform-provider-databricks/blob/main/catalog/resource_catalog.go#L48
Not sure if we want one struct or not. If we keep the `createPipelineRequestStruct`, let's put all methods together so it's more clear why this struct exists.

@edwardfeng-db What do you prefer?

Reply:
I think I prefer what Divyansh has here, because it is the same pattern with what I did with Clusters and Jobs.

I do agree that there are sharable contents in those `Aliases` functions tho, which we can extract and make more succinct